File: LC-Python-Solution/1011. Capacity To Ship Packages Within D Days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:

        l = max(weights)
        r = sum(weights)
        res = r

        def canShip(cap):
            ship = 1
            currCap = cap
            for w in weights:
                if currCap - w < 0:
                    ship+=1
                    currCap = cap
                currCap -=w
            return ship<=days

        while l<=r:
            cap = l + (r-l)//2
            if canShip(cap):
                res = min(cap,res)
                r = cap - 1            
            else:
                l = cap + 1
        
        return res






        

File: LC_JS-SOLUTIONS/1011. Capacity To Ship Packages Within D Days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function (weights, days) {
  let l = Math.max(...weights);
  let r = weights.reduce((acc, val) => acc + val, 0);
  let res = r;

  while (l <= r) {
    let cap = l + Math.floor((r - l) / 2);

    function isCapacity(cap) {
      let ship = 1;
      let capacity = cap;

      for (let i = 0; i < weights.length; i++) {
        if (capacity - weights[i] < 0) {
          ship += 1;
          capacity = cap;
        }
        capacity -= weights[i];
      }

      return ship <= days;
    }

    if (isCapacity(cap)) {
      res = Math.min(res, cap);
      r = cap - 1;
    } else {
      l = cap + 1;
    }
  }

  return res;
};
